syntax = "proto3";

package DATA;

message File {
    optional string name = 1;
    repeated string lines = 2;
}

message Version {
    optional string versionID = 1;
    optional string createdTime = 2;
    optional int32 versionNumber = 3; 
    optional int32 opsNumber = 4;//current operation number

    map<string, File> files= 5; //filename
}

message Project{
    optional string userID = 1;
    optional string projectID = 2;
    optional string projectName = 3;
    optional int32 currentVersion = 4; //Present version

    repeated Version versions = 5;
}

message Workspace {
    optional string userID = 1;

    map<string, Project> projects = 2; //name
}

message GitNote {
    map<string, Workspace> workspaces = 1; //userid
}